### PR TITLE
Clang build

### PR DESCRIPTION
**Pull request by klecki**
Fixes required to build witch clang. (tested with clang 6).
Example cmake command:
cmake -DCMAKE_CXX_COMPILER=clang++-6.0 -DCMAKE_C_COMPILER=clang-6.0 -DCUDA_HOST_COMPILER=/usr/bin/g++ ${SOURCE_DIR}

Requires #156

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -55,6 +55,11 @@ endif()
 # CXX flags
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -Wall -Werror -Wno-unused-variable -Wno-unused-function -fno-strict-aliasing -O2 -fPIC -fvisibility=hidden")
 
+if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated-register")
+endif()
+
+
```
Comment (JanuszL) [inline CMakeLists.txt]: Please add setting of gcc as a CUDA host compiler.

File: dali/image/generic_image.cc [hunk 2]
```diff
@@ -49,18 +49,19 @@ DALIError_t GetBMPImageDims(const uint8 *bmp, int size, int *h, int *w) {
     DALIError_t ret = DALIError;
```
Comment (JanuszL) [inline dali/image/generic_image.cc]: BMP fix is a separate patch. Please remove it from this PR.

Comment (klecki) [reply dali/image/generic_image.cc]: It is from #156 - but this is required by this PR to work.

File: dali/pipeline/data/types_test.cc [hunk 3]
```diff
@@ -44,14 +44,27 @@ TYPENAME_FUNC(float);
...
                          int16,
                          int32,
                          int64,
                          float16,
                          float,
                          double,
-                         bool> TestTypes;
+                         bool,
+                         std::vector<uint8>,
+                         std::array<std::vector<uint8>, 10>
```
Comment (cliffwoolley) [inline dali/pipeline/data/types_test.cc]: Why the magic number 10 here (and above)?

**Comment by klecki** [comment]
@cliffwoolley using clang can give better error checking (already discovered one bug) and compiler diagnostics (which can help during development). It could also allow to use additional tools, for example https://clang.llvm.org/docs/SourceBasedCodeCoverage.html
It is not meant to replace gcc but to aid the project.

File: dali/pipeline/data/types_test.cc (hunk 3 above)
Comment (klecki) [reply dali/pipeline/data/types_test.cc]: Done.

File: README.rst [hunk 4]
```diff
@@ -138,6 +138,13 @@ To build DALI with LMDB support:
    cmake -DBUILD_LMDB=ON ..
    make -j"$(nproc)"
 
+To build DALI using clang:
+
+.. code-block:: bash
+   
+   cmake -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_C_COMPILER=clang  ..
```
Comment (JanuszL) [inline README.rst]: Please add note that is it experimental and we are not testing that thoroughly and we recommend to use gcc for the productions builds.

File: dali/pipeline/data/types.h [hunk 5]
```diff
@@ -182,6 +182,13 @@ class DLL_PUBLIC TypeInfo {
   string name_;
 };
 
+template <typename T>
+struct DLL_PUBLIC TypeNameHelper {
```
Comment (JanuszL) [inline dali/pipeline/data/types.h]: Do those helpers need to be DLL_PUBLIC and public after all?

File: dali/pipeline/data/types_test.cc [hunk 6]
```diff
@@ -44,14 +48,27 @@ TYPENAME_FUNC(float);
...
 typedef ::testing::Types<uint8,
                          int16,
                          int32,
                          int64,
                          float16,
                          float,
                          double,
-                         bool> TestTypes;
+                         bool,
+                         std::vector<uint8>,
```
Comment (JanuszL) [inline dali/pipeline/data/types_test.cc]: Why we ned that?

**Comment by JanuszL** [comment]
@cliffwoolley -  as @klecki said, more compatibility won't do any harm and we can detect more errors during compilation (also if we need to support Tergra and some fancy compilers there it could only help). But we have no plan to use clang for production builds.

File: dali/pipeline/data/types_test.cc (hunk 6 above)
Comment (klecki) [reply dali/pipeline/data/types_test.cc]: To have better test coverage of this part - I wanted to make sure this change does not break anything.

File: README.rst (hunk 4 above)
Comment (klecki) [reply README.rst]: Done

File: dali/pipeline/data/types.h (hunk 5 above)
Comment (klecki) [reply dali/pipeline/data/types.h]: You're right, I forgot to remove this. Done.

**Comment by JanuszL** [comment]
@szalpal - it is possible, for that we need to add clang to the docker and provide some option which would select appropriate compiler. Then in our CI we could add some additional config for that.

**Comment by JanuszL** [comment]
Builds 35511.

**Pull request merged**